src/taxman.py

```diff
@@ -61,324 +78,606 @@ def __init__(self, book: Book, price_data: PriceData) -> None:
 
-    def _evaluate_taxation_per_coin(
-        self,
-        operations: list[transaction.Operation],
-    ) -> None:
-        """Evaluate the taxation for a list of operations per coin using
-        country specific functions.
+        else:
+            raise NotImplementedError
 
-        Args:
-            operations (list[transaction.Operation])
-        """
-        for coin, coin_operations in misc.group_by(operations, "coin").items():
-            coin_operations = transaction.sort_operations(coin_operations, ["utc_time"])
-            self.__evaluate_taxation(coin, coin_operations)
+    ###########################################################################
+    # General tax evaluation functions.
+    ###########################################################################
 
     def evaluate_taxation(self) -> None:
-        """Evaluate the taxation using country specific function."""
+        """Evaluate the taxation using country specific functions."""
         log.debug("Starting evaluation...")
 
         assert all(
             op.utc_time.year <= config.TAX_YEAR for op in self.book.operations
         ), "For tax evaluation, no operation should happen after the tax year."
 
-        if config.MULTI_DEPOT:
-            # Evaluate taxation separated by platforms and coins.
-            for _, operations in misc.group_by(
-                self.book.operations, "platform"
-            ).items():
-                self._evaluate_taxation_per_coin(operations)
-        else:
-            # Evaluate taxation separated by coins in a single virtual depot.
-            self._evaluate_taxation_per_coin(self.book.operations)
+        # Sort the operations by time.
+        operations = tr.sort_operations(self.book.operations, ["utc_time"])
+
+        # Evaluate the operations one by one.
+        # Difference between the config.MULTI_DEPOT and "single depot" method
+        # is done by keeping balances per platform and coin or only
+        # per coin (see self.balance).
+        for operation in operations:
+            self.__evaluate_taxation(operation)
+
+        # Evaluate the balance at deadline to calculate unrealized sells.
+        for balance in self._balances.values():
+            balance.sanity_check()
+
+            sold_coins = balance.remove_all()
+            for sc in sold_coins:
+                # Sum up the portfolio at deadline.
+                self.portfolio_at_deadline[sc.op.platform][sc.op.coin] += sc.sold
+
+                # "Sell" these coins which makes it possible to calculate the
+                # unrealized gain afterwards.
+                unrealized_sell = tr.Sell(
+                    utc_time=TAX_DEADLINE,
+                    platform=sc.op.platform,
+                    change=sc.sold,
+                    coin=sc.op.coin,
+                    line=[-1],
+                    file_path=Path(),
+                    fees=None,
+                )
+                self._evaluate_sell(
```

Comment:
This is done independent from `CALCULATE_UNREALIZED_GAINS`

Reply:
totally forgot about the variable :rofl: thanks